src/dc_tools.rs

```diff
@@ -670,13 +674,128 @@ pub fn remove_subject_prefix(last_subject: &str) -> String {
+        header.get(begin..begin + end)
+    })
+}
+
+pub(crate) fn parse_receive_header(header: &str) -> String {
+    let mut hop_info = String::from("Hop:\n");
+
+    if let Ok(date) = dateparse(header) {
+        let date_obj = Local.timestamp(date, 0);
+        hop_info += &format!("Date: {}\n", date_obj.to_rfc2822());
+    };
+
+    if let Some(from) = extract_address_from_receive_header(header, "from ") {
+        hop_info += &format!("From: {}\n", from.trim());
+    }
+
+    if let Some(by) = extract_address_from_receive_header(header, "by ") {
+        hop_info += &format!("By: {}\n", by.trim());
+    }
+    hop_info
+}
+
+/// parses "receive"-headers
+pub(crate) fn parse_receive_headers(headers: &Headers) -> String {
+    let headers = headers
+        .get_all_headers("Received")
+        .iter()
+        .rev()
+        .filter_map(|header_map_item| from_utf8(header_map_item.get_value_raw()).ok())
+        .enumerate()
+        .map(|(i, header_value)| (i + 1).to_string() + ". " + &parse_receive_header(header_value))
+        .collect::<Vec<_>>();
+
+    headers
+        .iter()
+        .map(|a| a.to_string())
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
 #[cfg(test)]
 mod tests {
     #![allow(clippy::indexing_slicing)]
 
     use super::*;
 
-    use crate::test_utils::TestContext;
+    use crate::{
+        config::Config, dc_receive_imf::dc_receive_imf, message::get_msg_info,
+        test_utils::TestContext,
+    };
+
+    #[test]
+    fn test_parse_receive_headers() {
+        let raw = include_bytes!("../test-data/message/mail_with_cc.txt");
+        let mail = mailparse::parse_mail(&raw[..]).unwrap();
+        let hop_info = parse_receive_headers(&mail.get_headers());
+        let expected = vec![
+            "1. Hop:",
+            // "Date: Sat, 14 Sep 2019 19:00:22 +0200\n",
+            "From: localhost",
+            "By: hq5.merlinux.eu",
+            "",
+            "2. Hop:",
+            // "Date: Sat, 14 Sep 2019 19:00:25 +0200\n",
+            "From: hq5.merlinux.eu",
+            "By: hq5.merlinux.eu",
+            "",
+        ];
+        // remove Date lines because they are not deterministic
+        let hop_info = hop_info
+            .split('\n')
+            .filter(|line| !line.starts_with("Date:"))
+            .collect::<Vec<_>>();
+        assert_eq!(hop_info, expected)
+    }
+
+    #[async_std::test]
+    async fn test_parse_receive_headers_integration() -> anyhow::Result<()> {
+        let t = TestContext::new_alice().await;
+        t.set_config(Config::ShowEmails, Some("2")).await?;
+        let raw = include_bytes!("../test-data/message/mail_with_cc.txt");
+        dc_receive_imf(&t, raw, "INBOX", 1, false).await.unwrap();
+        let g = t.get_last_msg().await;
+
+        let expected = r"State: Fresh
+
+hi
+
+Message-ID: 2dfdbde7@example.org
+Last seen as: INBOX/1
+1. Hop:
+From: localhost
+By: hq5.merlinux.eu
+
+2. Hop:
+From: hq5.merlinux.eu
+By: hq5.merlinux.eu
+";
```

Comment:
why such a multi-line format for each hop?  it actually makes it harder to read and also harder to machine-parse IMO.  
Each hop could just be a single line like "Hop: [From] [by] [date]" or maybe a different separator than " ".  As each mail has multiple hops typically this would be a lot less clutter for those cases one needs to go to Message-info (message from unverified sender to verified group etc.)

Reply:
On mobile devices the long lines would be wrapped around, anyway, so I think it's most human-readable in the current form.

We can still go with a single-line format of course as it's better machine-readable.

Reply:
Yup, we did go for a single-line format for better machine-readability.